### ISSUE TITLE
FR: relax JSON parsing of TokenResponse (and JsonWebToken)

### ISSUE THREAD

**Issue by wonderfly** [googleapis/google-oauth-java-client#62]
_From [michel.j...@e-office.com](https://code.google.com/u/108898467546456299511/) on November 28, 2012 00:30:25_

Version of google-oauth-java-client (e.g. 1.5.0-beta)?
1.12.0

Java environment (e.g. Java 6, Android 2.3, App Engine)?
AppEngine 1.7.3

Describe the problem.
When consuming Microsoft Office365 OAuth services, the accessToken that is retrieved through:

AuthorizationCodeTokenRequest . execute() as per the sample, I get the error:
number field formatted as a JSON string must use the @JsonString annotation How would you expect it to be fixed? In the JSONParser.java (from com.google.api.client.json)
I commented out  the Preconditions check, and now I have an accesstoken. And also the expires in has a value...

--------- start here -------

``` java
  case VALUE_STRING:

//        Preconditions.checkArgument(valueClass == null || !Number.class.isAssignableFrom(valueClass)
//            || fieldContext != null && fieldContext.getAnnotation(JsonString.class) != null,
//            "number field formatted as a JSON string must use the @JsonString annotation&#37;s",
//            contextString);
        // TODO(yanivi): "special" values like Double.POSITIVE_INFINITY?
        try {
          return Data.parsePrimitiveValue(valueType, getText());
        } catch (IllegalArgumentException e) {
          throw new IllegalArgumentException(contextString, e);
        }
```

_Original issue: http://code.google.com/p/google-oauth-java-client/issues/detail?id=62_

**Comment by lnikkila** [comment]
I’ve heard that this occurs with Dell’s Cloud Access Manager as well.

**Comment by dherben** [comment]
I'm having the same problem (with the expires_in formatted as a string) when authorizing against the ExactOnline accounting API.

I had managed to work around the problem for the initial token request by creating my own ExactOnlineResponseToken class (which maps the expires_in to a String) and executing the tokenRequest as such: 

```java
flow().newTokenRequest(authCode)
    .setRedirectUri(uri)
    .executeUnparsed()
    .parseAs(ExactOnlineTokenResponse.class)
    .toTokenResponse()
```

I feel this is better than having to tinker with the libraries themselves. Unfortunately, I still get the error when the Credential automatically tries to refresh the token, because that code is hardcoded in the Credential and I can't access it. 

Any plans on making the library more permissive in the near future? I guess I could perform my own checks before each request, proactively refreshing the token when necessary, although that feels hackish.

**Comment by Kah0ona** [comment]
@dherben I'm with you - this should be more permissive. There are so many services out there that are just a little bit wrongly implemented... 

I am also dealing with Exact Online at the moment. I for now went with the approach of @wonderfly, since I stopped reading there, and only after that found out about your post :-)

The unfortunate thing is that one only finds out about a bad upstream implementations when a complete integration of the google oauth workflow is already integrated in the application... 

Although technically ExactOnline should fix its implementation but it's so annoying to be dependent on companies to fix their proprietary shit :-1: 

So bottom line: a bit more permissiveness for all the non-string fields in the responses where a little instanceof check is put in place would help a lot of us out.

**Comment by PhoneixS** [comment]
I have also problems with office, and as a workaround I have changed the library like the code bellow. This is a workaround and in my opinion shouldn't be merged in the library. However the library must have more versatility and let you use a custom TokenResponse class like Apache OAuth library.

Workaround in TokenResponse.java:

```java
  /**
   * Lifetime in seconds of the access token (for example 3600 for an hour) or {@code null} for
   * none.
   */
  @Key("expires_in")
  private Object expiresInSeconds;

  /**
   * Returns the lifetime in seconds of the access token (for example 3600 for an hour) or
   * {@code null} for none.
   */
  public final Long getExpiresInSeconds() {
    if (expiresInSeconds instanceof Long) {
      return (Long)expiresInSeconds;
    }else if (expiresInSeconds instanceof Number) {
      expiresInSeconds = ((Number)expiresInSeconds).longValue();
      return (Long)expiresInSeconds;
    } else if (expiresInSeconds instanceof String) {
      expiresInSeconds = Long.parseLong((String)expiresInSeconds);
      return (Long)expiresInSeconds;
    }

    return null;
  }
```

**Comment by arobertn** [comment]
Hit this from Jenkins and worked around with the ravn fork.  I doubt Azure AD is going to change to comply with the standard anytime soon (it is Microsoft after all) so I hope the library itself can be made more permissive.  Log a warning (maybe at debug or info level) if that would make anyone feel better.  :-)

**Comment by chingor13** [comment]
Changing to a feature request. 

This issue is to support non-Google OAuth implementations that return different formatted JSON for the token response.

**Comment by chingor13** [comment]
Rather than relaxing the default types in `TokenResponse` to object to allow handling any possible permutation of responses, we will try to make it easier to use your own `TokenResponse` class by letting you redefine the JSON fields in a `TokenResponse` subclass and letting you specify your custom `TokenResponse` class when building a `TokenRequest`.

**Comment by Rinitha1** [comment]
Hi would like to know how do we modify the JSONParser claas or the TokenResponse class as they are in the jar file. Would be very helpful if anyone can share the steps.
